release-notes.txt: add 2022.01 release notes (#17568)
RIOT-OS/RIOT: release-notes.txt

============================
RIOT is a multi-threading operating system which enables soft real-time
capabilities and comes with support for a range of devices that are typically
found in the Internet of Things: 8-bit and 16-bit microcontrollers as well as
light-weight 32-bit processors.

RIOT is based on the following design principles: energy-efficiency, soft
real-time capabilities, small memory footprint, modularity, and uniform API
access, independent of the underlying hardware (with partial POSIX compliance).

RIOT is developed by an international open-source community which is
independent of specific vendors (e.g. similarly to the Linux community) and is
licensed with a non-viral copyleft license (LGPLv2.1), which allows indirect
business models around the free open-source software platform provided by RIOT.

About this release
==================

The 2022.01 release includes:

## List of notable changes here ##

436 pull requests, composed of 1159 commits, have been merged since the
last release, and 11 issues have been solved. 43 people contributed with
code in 88 days. 2475 files have been touched with 44925 (+) insertions and
14687 deletions (-).

> 

Reviewer: I would mention here: all the work regarding `rust`, that almost all `boards` where migrated to kconfig. The fact that with `ztimer64` all code can be migrated to `ztimer` from `xtimer`.

Author: Yup. Draft summary in the pipe. Stay tuned (and thanks for the suggestions @fjmolinas )

New features and changes
========================

Core
----

+ core/include/kernel_defines.h: add index_of() macro (#17142)
+ core: Add IS_CT_CONSTANT() (#17273)
+ core: add WITHOUT_PEDANTIC() and DECLARE_CONSTANT() (#17176)
* core: Warn about using irq_enable (#17433)

System Libraries
----------------

+ pkg/fff: Add fake functions framework package (#17076)
+ sys/architecture: add HAS_ALIGNMENT_OF() helper (#17177)
+ sys/arduino: add assertion to gpio handling (#17443)
+ sys/atomic_utils: add atomic_{load,store}_ptr() (#17256)
+ sys/include/event/periodic: add count (#17276)
+ sys/isrpipe: add isrpipe_write (#17336)
+ sys/sched_rr: Add a round robin scheduler module (#16126)
+ sys/stdio: add optional function stdio_available (#17446)
+ sys/test_utils/result_output: add sort to multiple output check (#17269)
+ sys/tsrb: add peek functions (#17266)
+ sys/tsrb: add tsrb_clear (#17337)
+ sys/ztimer: add 'ztimer_no_periph_rtt' (#17284)
+ sys/ztimer: add some best practices (#17370)
+ sys/ztimer: ztimer_now() add warning regarding comparing now() values (#17404)
+ usb/hid: Add HID report descriptor defines (#17242)
* drivers/periph_common: build periph module selectively (#17418)

> * event/timeout: remove forced ZTIMER_USEC dependency [alternative] (#16958)

Reviewer: ```suggestion
* event/timeout: remove forced ZTIMER_USEC dependency (#16958)
```

* riotboot_dfu: use ztimer instead of xtimer (#17051)
* riotboot_serial: enter bootloader mode by pin (#17248)
* schedstatistics: Convert to ztimer (#17216)
* stdio_rtt: Convert to ztimer (#17122)
* stdio_semihosting: Convert to ztimer (#17123)
* sys/can: migrate to ztimer (#17366)
* sys/cpp11-compat: use new/delete operators from sys/cpp_new_delete (#17475)
* sys/evtimer: use now returned by set (#17394)
* sys/fmt: make fmt_s32_dfp() string based (#17106)
* sys/Makefile.dep: fix event periodic dependency (#17318)
* sys/picolibc_syscalls_default: support new picolibc stdio globals (#17001)
* sys/random: default to musl LCG instead of TinyMT (#17188)
* sys/suit: adjust dependencies for CoAP transport (#17182)
* sys/test_utils/result_output/json: optional space after symbol (#17034)
* sys/turo: Allow default selection of json (#17039)
* sys/ztimer/xtimer2ztimer.cocci: extend and fix api replacements (#17234)
* sys/ztimer64: default select ztimer64_init (#17414)
* sys/ztimer64: make _del_entry_from_list() safe for uninit ztimer64_t (#17368)
* sys/ztimer: remove double ztimer indirection (#17272)
* sys/ztimer: ztimer_remove report success (#17408)

> * usbdev/USBUS: deprecate transfer failure reporting report (#17046)

Reviewer: I would move this to deprecations...

* pkg/lwip: Fix dualstack build when only using 6lowpan (#17174)
* pkg/openthread: bump version to 20200818 + adapt to use CMake to
  configure the build (#16948)
* pkg/openthread: migrate to ztimer (#17120)
* pkg/paho-mqtt: migrate to ztimer (#17143)
* pkg/semtech-loramac: fix get/set dr command (#17117)
* sys/evtimer,ztimer: do not depend on ztimer_now64 (#17357)
* sys/net/gnrc: fix compilation with -Wcast-align (#17157)
* sys/shell/gnrc_netif: Allow 'ifconfig help' (#17006)
* test/ieee802154_hal: general cleanup and cosmetic fixes (#16913)
* tests/lwip_sock: Take lock before calling etharp (#17145)
* Update tiny-asn1 package (#16936)

Packages
--------

+ pkg/lz4: add support for LZ4 compression/decompression (#17178)
+ pkg/uzlib: add support for zlib compression/decompression (#17179)
+ pkg: add support for Elk Tiny Javascript engine (#16887)
+ pkg: add support for FreeRTOS coreJSON library (#16997)
* hacl: Switch git address to RIOT-OS-pkgmirror (#16941)
* make: support package mirrors (#16927)
* pkg/emlearn: bump to 0.12 (#16937)
* pkg/jerryscript: bump to 2.4.0 (#16938)
* pkg/mynewt-core/patches: silence cast-align (#17397)
* pkg/qr-code-generator: bump to 1.7.0 (#16939)

> * pkg/uwb-core: fix wrong header include (#17345)
> * pkg/uwb-core: multiple cleanups (#17264)

Reviewer: ```suggestion
* pkg/uwb-core: fix wrong header include (#17345)
* pkg/uwb-core: enable frame filtering, allow usage o common event thread (#17264)
```

* tests/periph_timer_periodic: spice up test (#17388)
* tests/unittests/tests-base64: enlarge test buffer for worst case (#17292)

API Changes
-----------

* core/kernel_defines.h: drop ALIGN_OF() (#17267)
* core/kernel_defines: drop BUILD_BUG_ON() (#17268)
* core/msg: make msg_avail() return 0 on no queue (#17262)
* cpu: call cpu_init() from startup code, not from board (#16055)
* drivers/periph_i2c: let i2c_acquire return void (#17275)
* sys/ztimer: ztimer_set() return the now value (#17385)
* USB: refactor to xmit-based API (#17064)

Uncategorized
-------------
+ LOSTANDFOUND.md: add boards/fox (#17041)
+ release-notes.txt: add 2021.10 release notes [backport 2021.10] (#17070)
+ SECURITY.md: add gpg key information (#17189)
+ SECURITY.md: Add key link (#17223)
* .github/ISSUE_TEMPLATE: remove security_bug.md (#17167)
* CITATION.cff: Initial import (#17028)
* CODING_CONVENTIONS_C++.md: Change space after negation (#17071)
* Remove merge chance shield (deprecated) (#17042)

And 79 minor changes.

Deprecations
============

Deprecations (2)
----------------

* cpu/mips: deprecate mips cpu (#17304)
* sys/include/xtimer.h: deprecate nanosleep (#17296)

> 

Reviewer 2: ```suggestion
* sys/include/evtimer.h: deprecate evtimer_now_min (#17357)

```

Reviewer 2: @emmanuelsearch this

RIOT-2022.01 - Release Notes
============================
RIOT is a multi-threading operating system which enables soft real-time
capabilities and comes with support for a range of devices that are typically
found in the Internet of Things: 8-bit and 16-bit microcontrollers as well as
light-weight 32-bit processors.

RIOT is based on the following design principles: energy-efficiency, soft
real-time capabilities, small memory footprint, modularity, and uniform API
access, independent of the underlying hardware (with partial POSIX compliance).

RIOT is developed by an international open-source community which is
independent of specific vendors (e.g. similarly to the Linux community) and is
licensed with a non-viral copyleft license (LGPLv2.1), which allows indirect
business models around the free open-source software platform provided by RIOT.

About this release
==================

The 2022.01 release brings support for new packages, new boards and drivers,

> as well as several pkg upgrades and bug fixes. Among others, this release

Reviewer 3: One last nit:

```suggestion
as well as several package upgrades and bug fixes. Among others, this release
```

improves the RIOT codebase with:

Rust support
------------

RIOT now offers convenient integration of Rust application programming, with

> wrappers interfacing to RIOT modules such as timers, network stack;

Reviewer: ```suggestion
wrappers interfacing to RIOT modules such as timers, network stack.
```

Improved timers
---------------

Many modules have migrated from using xtimer to our more power-efficient
alternative ztimer. Also, ztimer64 was introduced, which paves the way

> for deprecating xtimer soon;

Reviewer: ```suggestion
for deprecating xtimer soon.
```

> alternative ztimer. Also, ztimer64 was introduced, which paves the way
> for deprecating xtimer soon;

Reviewer: ```suggestion
alternative ztimer. Also, ztimer64 was introduced, which allows for
a drop-in replacement for all of xtimer features.
```

* cpu/native/include: other stacksizes based on default (#16956)
* cpu/native: migrate periph_rtc to ztimer (#17125)
* cpu/qn908x: migrate ADC periph to ztimer (#17321)
* cpu/sam0_common: SPI: don't perform DMA transfer for small buffers (#16926)
* cpu/sam0_common: uart: set oversampling based on baud rate (#17186)
* cpu/stm32/include/cpu_conf_stm32_common.h: fix typo in macro (#17187)
* cpu/stm32: bump cmsis repositories to latest releases + extend CPU
  models kconfig (#17293)
* cpu/stm32: duplicated ADC devices definitions (#16968)
* cpu/stm32: migrate adc and eth periphs to ztimer (#17320)
* cpu/stm32: split periph_cpu.h in sub headers (#17441)
* cpu: do not auto-select ztimer_periph_rtt for sam0, fe310 (#17395)
* drivers/periph: use uint_fast8_t as default type (#17191)
* esp8266: Download Espressif RTOS SDK as a new RIOT PKG (#16425)
* periph/flashpage: extend API (#16972)
* periph/flashpage: fix wrong doxygen group (#17202)
* stm32/usbdev: Use ztimer instead of xtimer (#17047)
* cpu/esp: compile optional modules in CI (#17314)
* cpu/stm32/wl: initial periph_adc implementation (#17183)
* cpu/stm32: enable USB OTG FS on high end L4 (#17302)

> 

Reviewer 3: ```suggestion
```

+ netdev: add netdev_register_signal (#16922)
+ drivers/ft5x06: add support for touch panel controller (#17448)
+ drivers/stmpe811: add spi mode (#17088)
+ drivers/dose: make use of UART collision detection feature (#16681)
* driver/at30tse75x: port to ztimer_usec (#17137)
* drivers/ads101x: migrate to ztimer (#17102)
* drivers/ata8520e: migrate to ztimer (#17105)
* drivers/atwinc15x0: migrate to ztimer (#17381)
* drivers/bmp180: migrate to ztimer (#17101)
* drivers/dose: calculate timeout based on symbol rate (#16768)
* drivers/dsp0401: migrate to ztimer (#17110)
* drivers/mhz19: migrate to ztimer (#17308)
* drivers/rn2xx3: migrate to ztimer (#17112)
* drivers/sdcard_spi: convert to ztimer_usec (#17361)
* drivers/stmpe811: improve interrupt callback management (#17487)
* drivers/xbee: migrate to ztimer_msec (#17114)
* drivers: migrate mpu9x50 and si70xx to ztimer (#17358)
* drivers: migrate some drivers to ztimer (#17315)
* drivers: remove some useless xtimer includes (#17303)
* drivers: several cleanups related to xtimer + some ztimer migration (#17375)
* makefiles: remove PCF857x pseudomodule definitions (#17343)
* drivers/periph_common: build periph module selectively (#17418)

> 

Reviewer 3: ```suggestion
```

wrappers interfacing to RIOT modules such as timers, network stack;

Improved timers
---------------

Many modules have migrated from using xtimer to our more power-efficient
alternative ztimer. Also, ztimer64 was introduced, which brings paves the way
for deprecating xtimer soon;

Kconfig support
---------------

Kconfig modeling coverage is significantly extended. Nearly all supported
boards are now modeled with Kconfig.

Summary:
--------

436 pull requests, composed of 1159 commits, have been merged since the
last release, and 11 issues have been solved. 43 people contributed with
code in 88 days. 2475 files have been touched with 44925 (+) insertions and
14687 deletions (-).

Notations used below
====================

    + means new feature/item
    * means modified feature/item
    - means removed feature/item

New features and changes
========================

Core
----

+ core/include/kernel_defines.h: add index_of() macro (#17142)
+ core: Add IS_CT_CONSTANT() (#17273)
+ core: add WITHOUT_PEDANTIC() and DECLARE_CONSTANT() (#17176)
* core: Warn about using irq_enable (#17433)

System Libraries
----------------

> + pkg/fff: Add fake functions framework package (#17076)

Reviewer 3: Should rather be listed in the pkg section

Reviewer 3: Or even "Testing"

+ sys/architecture: add HAS_ALIGNMENT_OF() helper (#17177)
+ sys/arduino: add assertion to gpio handling (#17443)
+ sys/atomic_utils: add atomic_{load,store}_ptr() (#17256)
+ sys/include/event/periodic: add count (#17276)
+ sys/isrpipe: add isrpipe_write (#17336)
+ sys/sched_rr: Add a round robin scheduler module (#16126)
+ sys/stdio: add optional function stdio_available (#17446)
+ sys/test_utils/result_output: add sort to multiple output check (#17269)
+ sys/tsrb: add peek functions (#17266)
+ sys/tsrb: add tsrb_clear (#17337)
+ sys/ztimer: add 'ztimer_no_periph_rtt' (#17284)
+ sys/ztimer: add some best practices (#17370)
+ sys/ztimer: ztimer_now() add warning regarding comparing now() values (#17404)
+ usb/hid: Add HID report descriptor defines (#17242)

> * drivers/periph_common: build periph module selectively (#17418)

Reviewer 3: Should go in the drivers section

* event/timeout: remove forced ZTIMER_USEC dependency (#16958)

> * examples/gnrc_networking: cleanup and static function declarations in

Reviewer 3: Should go in the examples section

  udp.c (#17384)
* pkg/arduino_sdi_12: support the remote-revb board (#17119)

> * pkg/fatfs: bump version to r0.14b (#16412)

Reviewer 3: Both should go to pkg section

* sys/picolibc_syscalls_default: support new picolibc stdio globals (#17001)
* sys/random: default to musl LCG instead of TinyMT (#17188)
* sys/suit: adjust dependencies for CoAP transport (#17182)
* sys/test_utils/result_output/json: optional space after symbol (#17034)
* sys/turo: Allow default selection of json (#17039)
* sys/ztimer/xtimer2ztimer.cocci: extend and fix api replacements (#17234)
* sys/ztimer64: default select ztimer64_init (#17414)
* sys/ztimer64: make _del_entry_from_list() safe for uninit ztimer64_t (#17368)
* sys/ztimer: remove double ztimer indirection (#17272)
* sys/ztimer: ztimer_remove report success (#17408)
* usbus/cdc_acm: Return stall on line coding not supported (#17085)
* usbus/hid: fix ep_out readiness (#17230)
* usbus/hid: fix ep_out readiness (#17245)
* usbus: Bind extra USB config (#17380)
* usbus: check received setup request data amount (#17203)

Networking
----------

+ gnrc_pktbuf_cmd: add od dependency with gnrc_pktbuf_static (#17228)
+ sys/net/dhcpv6: Add stateless DHCPv6 (#16731)
+ sys/net: Add ipv4/ipv6 pseudomodules (#16966)
+ sys/shell/sc_nimble_netif: add ping subcommand (#16539)

> * doc: fix missing or non matching arguments in Doxygen documentation (#17461)

Reviewer 3: Move to doc section

* dsm: rename DTLS_PEER_MAX and make it dependent on
  CONFIG_DTLS_PEER_MAX (#16967)
* gcoap_dtls: destroy session in _tl_send only on connection errors (#16963)
* gnrc_dhcpv6_client / uhcp: delay router advertisements until prefix
  was received (#16755)
* gnrc_gomach: adapt duty recording parameters' namings. (#8954)
* gnrc_netif: move netdev initialization to ops->init (#16918)
* lwip: bump to v2.1.3 (#16774)
* net/gcoap: port to ztimer (#17141)
* nimble/statconn: use errno return values (#17344)
* pkg/lwip: Don't control IPv6 via IPv4 flag (#17175)
* pkg/lwip: Fix dualstack build when only using 6lowpan (#17174)
* pkg/openthread: bump version to 20200818 + adapt to use CMake to
  configure the build (#16948)
* pkg/openthread: migrate to ztimer (#17120)
* pkg/paho-mqtt: migrate to ztimer (#17143)
* pkg/semtech-loramac: fix get/set dr command (#17117)
* sys/evtimer,ztimer: do not depend on ztimer_now64 (#17357)
* sys/net/gnrc: fix compilation with -Wcast-align (#17157)
* sys/shell/gnrc_netif: Allow 'ifconfig help' (#17006)
* test/ieee802154_hal: general cleanup and cosmetic fixes (#16913)
* tests/lwip_sock: Take lock before calling etharp (#17145)

> * Update tiny-asn1 package (#16936)

Reviewer 3: Should move to pkg section

Packages
--------

+ pkg/lz4: add support for LZ4 compression/decompression (#17178)
+ pkg/uzlib: add support for zlib compression/decompression (#17179)
+ pkg: add support for Elk Tiny Javascript engine (#16887)
+ pkg: add support for FreeRTOS coreJSON library (#16997)
* hacl: Switch git address to RIOT-OS-pkgmirror (#16941)
* make: support package mirrors (#16927)
* pkg/emlearn: bump to 0.12 (#16937)
* pkg/jerryscript: bump to 2.4.0 (#16938)
* pkg/mynewt-core/patches: silence cast-align (#17397)
* pkg/qr-code-generator: bump to 1.7.0 (#16939)
* pkg/uwb-core: fix wrong header include (#17345)
* pkg/uwb-core: enable frame filtering, allow usage of common event thread (#17264)
* pkg/wakaama: use ztimer (#17103)
* test/pkg_u8g2: increase main stacksize for SDL (#16954)

> * tests/pkg_fatfs{_vfs}: drop whitelist (#17478)

Reviewer 3: Should go to the Testing section

Boards
------

+ boards/feather-m0: add arduino feature (#17401)
+ boards/iotlab-m3: add MTD definition (#17339)
+ boards/nrf9160dk: add riotboot support (#17226)
+ boards/nucleo-g431rb: added PWM configuration (#17030)
+ boards/p-l496g-cell02: add riotboot feature (#17463)
+ boards/stm32f469i-disco: Add DAC (#17214)
+ boards/stm32g031-disco,examples,tests: Added STM32G031-DISCO board (#16959)
+ boards: add board support for Alientek Pandora (#17259)
+ boards: add support for stm32f746g-disco (#17198)
+ cpu/esp32: Add openocd programming support (#16911)
+ cpu/stm32: add driver for the LTDC peripheral (#17437)
+ cpu/stm32: add new usbdev FS support (#17281)

> + cpu/stm32: add support for U5 family (#17410)

Reviewer 3: Should move to the CPU section

+ drivers/ft5x06: add support for touch panel controller (#17448)

> + drivers/stmpe811: add spi mode (#17088)

Reviewer 3: Should move to the Drivers section

> + makefiles/cflags.inc.mk: Add -Wcast-align (#14955)

Reviewer 3: Should move to the Build system section

> + pkg/arduino_sdi_12: add SDI-12 for Arduino as package (#16587)

Reviewer 3: Should move to Packages section

+ sys/arduino: add Serial over stdio support (#17447)

> + sys: introduce sysclk function to retrieve core clock frequency (#17342)

Reviewer 3: Should move to System section

* boards/common/microbit: fix doxygen grouping (#17440)
* boards/esp* and cpu/esp: feature definition cleanup (#17438)
* boards/lora-e5-dev: enable 3.3V and 5V output by default (#17158)
* boards/lora-e5-dev: fix default DARWIN port (#17373)
* boards/stm32f469i-disco: Create  board (#16898)
* boards/{calliope-mini,microbit*}: factorize common microbit module,
  use ztimer (#17124)
* boards: cpu: replace GPIO_UNDEF by SPI_CS_UNDEF in SPI configuration (#17468)
* boards: model nrf52 boards (#17299)
* boards: unify user button defines (#17428)
* cpu/esp: compile optional modules in CI (#17314)
* cpu/stm32/wl: initial periph_adc implementation (#17183)

> * cpu/stm32: enable USB OTG FS on high end L4 (#17302)

Reviewer 3: Should move to CPU

> * socket_zep: port to radio HAL (#16932)

Reviewer 3: Maybe move to Networking ?

* tools/elf2uf2: addition of new PROGRAMMER for RPi-pico (#17348)

CPU
---

+ Add some Rust building infrastructure and example (#16274)
+ cpu/esp32: add clear bus during init to the I2C software
  implementation (#17346)
+ cpu/nrf9160: add periph_flashpage support (#17225)
+ cpu/stm32: Add hardening changes to stm32 (#16979)

> + drivers/dose: make use of UART collision detection feature (#16681)

Reviewer 3: Move to Drivers

* cpu/kinetis/include: fix xtimer backend timer selection (#17393)
* cpu/native/include: other stacksizes based on default (#16956)
* cpu/native: migrate periph_rtc to ztimer (#17125)
* cpu/qn908x: migrate ADC periph to ztimer (#17321)
* cpu/sam0_common: SPI: don't perform DMA transfer for small buffers (#16926)
* cpu/sam0_common: uart: set oversampling based on baud rate (#17186)
* cpu/stm32/include/cpu_conf_stm32_common.h: fix typo in macro (#17187)
* cpu/stm32: bump cmsis repositories to latest releases + extend CPU
  models kconfig (#17293)
* cpu/stm32: duplicated ADC devices definitions (#16968)
* cpu/stm32: migrate adc and eth periphs to ztimer (#17320)
* cpu/stm32: split periph_cpu.h in sub headers (#17441)
* cpu: do not auto-select ztimer_periph_rtt for sam0, fe310 (#17395)
* drivers/periph: use uint_fast8_t as default type (#17191)
* esp8266: Download Espressif RTOS SDK as a new RIOT PKG (#16425)
* Fix some cppcheck errors (#17271)
* ieee802154/hal: adapt to latest changes of #13943 (#16946)
* periph/flashpage: extend API (#16972)
* periph/flashpage: fix wrong doxygen group (#17202)
* stm32/usbdev: Use ztimer instead of xtimer (#17047)

> * sys/ztimer64: initial PR (#16928)

Reviewer 3: Move to Sys

* treewide: Address uncrustify suggestions (#16737)

> * usb(dev|bus): Fix static-check issues (#17135)

Reviewer 3: Maybe move to Testing (sounds more like CI fixes)

+ drivers: add PCF857X I2C I/O expander driver (#10430)
+ drivers: only build saul interface if saul module is used + add
  saul_drivers test application (#17470)
+ netdev: add netdev_register_signal (#16922)
* driver/at30tse75x: port to ztimer_usec (#17137)
* drivers/ads101x: migrate to ztimer (#17102)
* drivers/ata8520e: migrate to ztimer (#17105)
* drivers/atwinc15x0: migrate to ztimer (#17381)
* drivers/bmp180: migrate to ztimer (#17101)
* drivers/dose: calculate timeout based on symbol rate (#16768)
* drivers/dsp0401: migrate to ztimer (#17110)
* drivers/mhz19: migrate to ztimer (#17308)
* drivers/rn2xx3: migrate to ztimer (#17112)
* drivers/sdcard_spi: convert to ztimer_usec (#17361)
* drivers/stmpe811: improve interrupt callback management (#17487)
* drivers/xbee: migrate to ztimer_msec (#17114)
* drivers: migrate mpu9x50 and si70xx to ztimer (#17358)
* drivers: migrate some drivers to ztimer (#17315)
* drivers: remove some useless xtimer includes (#17303)
* drivers: several cleanups related to xtimer + some ztimer migration (#17375)
* makefiles: remove PCF857x pseudomodule definitions (#17343)
* pkg/lvgl: use ztimer instead of xtimer (#17031)

> * pkg: migrate some packages to ztimer (#17309)

Reviewer 3: Move to packages

found in the Internet of Things: 8-bit and 16-bit microcontrollers as well as
light-weight 32-bit processors.

RIOT is based on the following design principles: energy-efficiency, soft
real-time capabilities, small memory footprint, modularity, and uniform API
access, independent of the underlying hardware (with partial POSIX compliance).

RIOT is developed by an international open-source community which is
independent of specific vendors (e.g. similarly to the Linux community) and is
licensed with a non-viral copyleft license (LGPLv2.1), which allows indirect
business models around the free open-source software platform provided by RIOT.

About this release
==================

The 2022.01 release brings support for new packages, new boards and drivers,
as well as several pkg upgrades and bug fixes. Among others, this release
improves the RIOT codebase with:

Rust support
------------

RIOT now offers convenient integration of Rust application programming, with
wrappers interfacing to RIOT modules such as timers, network stack;

Improved timers
---------------

Many modules have migrated from using xtimer to our more power-efficient

> alternative ztimer. Also, ztimer64 was introduced, which brings paves the way

Reviewer 3: ```suggestion
alternative ztimer. Also, ztimer64 was introduced, which paves the way
```

+ boards: add board support for Alientek Pandora (#17259)
+ boards: add support for stm32f746g-disco (#17198)
* boards/common/microbit: fix doxygen grouping (#17440)
* boards/esp* and cpu/esp: feature definition cleanup (#17438)
* boards/lora-e5-dev: enable 3.3V and 5V output by default (#17158)
* boards/lora-e5-dev: fix default DARWIN port (#17373)
* boards/stm32f469i-disco: Create  board (#16898)
* boards/{calliope-mini,microbit*}: factorize common microbit module,
  use ztimer (#17124)
* boards: cpu: replace GPIO_UNDEF by SPI_CS_UNDEF in SPI configuration (#17468)
* boards: model nrf52 boards (#17299)
* boards: unify user button defines (#17428)
* tools/elf2uf2: addition of new PROGRAMMER for RPi-pico (#17348)

CPU
---

+ Add some Rust building infrastructure and example (#16274)
+ cpu/esp32: add clear bus during init to the I2C software
  implementation (#17346)
+ cpu/nrf9160: add periph_flashpage support (#17225)
+ cpu/stm32: Add hardening changes to stm32 (#16979)
+ nimble_scanner: rework to enable scanning BLE5 PHYs and extended
  advertisements (#16843)
+ nrf5x_common: Add qdec peripheral implementation (#17201)

> + sys/new_delete: add malloc/free based new/delete implementation (#17464)

Reviewer 3: Should be moved to sys (missed that one yesterday)

+ cpu/esp32: Add openocd programming support (#16911)
+ cpu/stm32: add driver for the LTDC peripheral (#17437)
+ cpu/stm32: add new usbdev FS support (#17281)
+ cpu/stm32: add support for U5 family (#17410)
* boards/common/esp8266: Set RTS and DTR to 0 on pyterm. (#16305)
* cc2538_rf: remove RX Busy detection in request_set_trx_state (#16821)

> * core: make SCHED_TEST_STACK boolean and default to 1 with DEVELHELP (#17132)

Reviewer 3: Should be moved to core

* cpu/esp32: make CI happy (#17221)
* cpu/esp32: place freertos and periph in IRAM (#17087)
* cpu/esp32: support CPU clocks 2 MHz and 40 MHz (#17413)
* cpu/esp: change dynamic SSID option handling (#17415)
* cpu/esp: migrate to ztimer (#17386)
* cpu/kinetis/include: fix xtimer backend timer selection (#17393)
* cpu/native/include: other stacksizes based on default (#16956)
* cpu/native: migrate periph_rtc to ztimer (#17125)
* cpu/qn908x: migrate ADC periph to ztimer (#17321)
* cpu/sam0_common: SPI: don't perform DMA transfer for small buffers (#16926)
* cpu/sam0_common: uart: set oversampling based on baud rate (#17186)
* cpu/stm32/include/cpu_conf_stm32_common.h: fix typo in macro (#17187)
* cpu/stm32: bump cmsis repositories to latest releases + extend CPU
  models kconfig (#17293)
* cpu/stm32: duplicated ADC devices definitions (#16968)
* cpu/stm32: migrate adc and eth periphs to ztimer (#17320)
* cpu/stm32: split periph_cpu.h in sub headers (#17441)
* cpu: do not auto-select ztimer_periph_rtt for sam0, fe310 (#17395)
* drivers/periph: use uint_fast8_t as default type (#17191)
* esp8266: Download Espressif RTOS SDK as a new RIOT PKG (#16425)

> * Fix some cppcheck errors (#17271)

Reviewer 3: Should be moved to Testing

> * ieee802154/hal: adapt to latest changes of #13943 (#16946)

Reviewer 3: Should be moved to Networking

+ cpu/nrf9160: add Kconfig dependencies (#17291)
+ cpu/riscv: model Kconfig (#16909)
+ makefiles/kconfig: add board & CPU config variables (#17376)
+ pkg/uwb*: add Kconfig dependency modelling (#16780)
+ sys/trace: migrate to ztimer and add Kconfig support (#17319)
+ .murdock: Add modules/pkgs diff of make/kconfig (#17094)
+ .murdock: Add nucleo boards to kconfig test (#16845)
+ .murdock: Add same54-xpro to kconfig tests (#17255)
* .murdock: Kconfig test apps for all boards (#17402)
* boards/adafruit-itsybitsy-m4: Model kconfig (#17331)
* boards/feather-m0*: base board definition in Kconfig (#17445)
* boards/samd21-based: model Kconfig (#17355)
* boards/saml21-based: model kconfig (#17270)
* boards/stm32-based: model Kconfig (#17374)
* cpu/esp{32,8266}: model kconfig (#17232)
* cpu/kinetis: model kconfig (#17235)
* cpu/lpc23xx: model kconfig (#17279)
* cpu/nrf51: model kconfig (#17274)
* cpu/qn908x: model kconfig (#17277)
* cpu/{cc26xx, cc13xx}: model kconfig (#17236)
* drivers/sx126x/Kconfig: rework model selection (#17289)
* sys/benchmark/kconfig: cleanup ztimer dependency (#17317)
* sys/log: model Kconfig (#17286)
* sys/vfs: model Kconfig (#17287)

> 

Reviewer 3: ```suggestion
```

* USB: refactor to xmit-based API (#17064)

Deprecations
============

Deprecations (2)
----------------

* cpu/mips: deprecate mips cpu (#17304)
* sys/include/xtimer.h: deprecate nanosleep (#17296)
* sys/include/evtimer.h: deprecate evtimer_now_min (#17357)
* usbdev/USBUS: deprecate transfer failure reporting report (#17046)

Removals (9)
------------

- boards/fox: remove BOARD (#17021)
- dist/iotlab: remove deprecated auto-ssh backward compatibility (#17476)
- drivers/ethos: remove deprecated USE_ETHOS_FOR_STDIO define (#17499)
- make: remove deprecation warnings raised with flasher tools old vars (#17490)
- pkg/libcoap: remove (#17163)
- sys/asymcute: remove deprecated CONFIG_ASYMCUTE_BUFSIZE_EXP (#17481)
- sys/crypto: remove deprecated CIPHER_AES_128 (#17480)
- sys/saul_reg: remove deprecated saul_reg_rm function (#17489)
- sys/shell: remove deprecated SHELL_NO_{ECHO,PROMPT} defines (#17484)

Bug fixes (66)
==============

* cpu/native: add `-no-pie` to LINKFLAGS (#16186)
* boards/adafruit-clue: use internal RC oscillator (#17069)
* boards/feather-m0-*: fix Arduino feature in Kconfig (#17444)
* boards/nrf9160dk: fix LED macros (#17328)

> * boards/nucleo-g07xrb: fix or exclude some doxygen warnings (#17079)
> * boards/nucleo-g07xrb: fix or exclude some doxygen warnings [backport
>   2021.10] (#17084)

Reviewer 3: There's a duplicate here. I would drop the 2021.10 backport.
```suggestion
* boards/nucleo-g07xrb: fix or exclude some doxygen warnings (#17079)
```

Reviewer 3: Still not addressed

Reviewer 3: Still not addressed (and a couple below as well)...
